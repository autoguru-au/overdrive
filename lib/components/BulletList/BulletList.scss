.root {
	list-style-position: inside;
}

.firstOccurrence {
	padding-left: var(--global--space--6);
}
