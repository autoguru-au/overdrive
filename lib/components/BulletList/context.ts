import { Context, createContext } from 'react';

export type BulletType = 'circle' | 'square' | 'disc';

export const bulletMap: BulletType[] = ['circle', 'square', 'disc'];

export const BulletListContext: Context<number> = createContext(-1);
