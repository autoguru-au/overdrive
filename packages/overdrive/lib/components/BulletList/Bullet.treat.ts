import { style, styleMap } from 'treat';

export const root = styleMap((theme) => ({
	default: {
		display: 'grid',
		gridTemplateColumns: 'min-content 1fr',
		gridGap: theme.space['4'],
		marginBottom: theme.space['2'],

		':last-of-type': {
			marginBottom: 0,
		},

		':before': {
			display: 'block',
			alignSelf: 'flex-start',
			width: theme.space['2'],
			height: theme.space['2'],
			marginTop: `calc((1.6em - ${theme.space['2']}) / 2)`,
			content: "' '",
			backgroundColor: theme.typography.colour.dark,
		},
	},
	circle: {
		':before': {
			borderRadius: '50%',
		},
	},
	disc: {
		':before': {
			border: `2px solid ${theme.typography.colour.dark}`,
			borderRadius: '50%',
			backgroundColor: 'transparent',
		},
	},
	square: {
		':before': {
			border: `2px solid ${theme.typography.colour.dark}`,
			backgroundColor: 'transparent',
		},
	},
}));

export const noDot = style({
	':before': {
		visibility: 'hidden',
	},
});
