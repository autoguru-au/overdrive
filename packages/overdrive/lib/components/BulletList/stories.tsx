import * as React from 'react';
import { Bullet, BulletList } from '.';

import { Text } from '../Text';

export default {
	title: 'Foundation|List/BulletList',
	component: BulletList,
	decorators: [
		(story) => (
			<div style={{ maxWidth: '500px', width: '100%' }}>{story()}</div>
		),
	],
};

export const standard = () => (
	<BulletList>
		<Bullet>
			<Text>
				For some, it could be argued that air-conditioning is the most
				important feature of a car, excluding all the mechanical bits
				that make it go, stop and turn.
			</Text>
		</Bullet>
		<Bullet>
			<Text>
				There are many components that have to work together to ensure
				the air-conditioning system is working correctly and, with so
				many contributing parts, itâ€™s clear the system should only be
				serviced or repaired by a qualified technician.
			</Text>
		</Bullet>
	</BulletList>
);
export const nested = () => (
	<BulletList>
		<Bullet>
			<Text>
				For some, it could be argued that air-conditioning is the most
				important feature of a car, excluding all the mechanical bits
				that make it go, stop and turn.
			</Text>
		</Bullet>
		<Bullet>
			<BulletList>
				<Bullet>
					<Text>
						For some, it could be argued that air-conditioning is
						the most important feature of a car, excluding all the
						mechanical bits that make it go, stop and turn.
					</Text>
				</Bullet>
				<Bullet>
					<BulletList>
						<Bullet>
							<Text>
								For some, it could be argued that
								air-conditioning is the most important feature
								of a car, excluding all the mechanical bits that
								make it go, stop and turn.
							</Text>
						</Bullet>
						<Bullet>
							<BulletList>
								<Bullet>
									<Text>
										For some, it could be argued that
										air-conditioning is the most important
										feature of a car, excluding all the
										mechanical bits that make it go, stop
										and turn.
									</Text>
								</Bullet>
							</BulletList>
						</Bullet>
					</BulletList>
				</Bullet>
			</BulletList>
		</Bullet>
		<Bullet>
			<Text>
				For some, it could be argued that air-conditioning is the most
				important feature of a car, excluding all the mechanical bits
				that make it go, stop and turn.
			</Text>
		</Bullet>
	</BulletList>
);
