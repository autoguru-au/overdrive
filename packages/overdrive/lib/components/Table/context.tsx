import { invariant } from '@autoguru/utilities';
import type { FunctionComponent } from 'react';
import * as React from 'react';
import { createContext, ReactNode, useContext, useMemo } from 'react';

import { Tokens } from '../../themes/tokens';

export interface TableContext {
	padding?: keyof Tokens['space'];
	stickyHead?: boolean;
	children?: ReactNode;
}

const tableContext = createContext<TableContext | null>(null);

export const TableContextProvider: FunctionComponent<TableContext> = ({
	padding,
	stickyHead,
	children,
}) => {
	const value = useMemo(
		() => ({
			padding,
			stickyHead,
		}),
		[padding, stickyHead],
	);
	return (
		<tableContext.Provider value={value}>{children}</tableContext.Provider>
	);
};

export const useTableContext = () => {
	const ctx = useContext(tableContext);

	invariant(
		ctx !== null,
		"Make sure you've got a <Table /> component in your tree",
	);

	return ctx;
};
