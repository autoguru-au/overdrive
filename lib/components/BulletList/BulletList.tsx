import clsx from 'clsx';
import * as React from 'react';
import { FunctionComponent, useContext } from 'react';
import { useStyles } from 'react-treat';

import { Box } from '../Box';
import * as styleRefs from './BulletList.treat';
import { BulletListContext, bulletMap } from './context';

interface Props {
	className?: string;
}

export const BulletList: FunctionComponent<Props> = ({
	children,
	className,
}) => {
	const stack = useContext(BulletListContext);
	const styles = useStyles(styleRefs);

	return (
		<Box
			is="ul"
			className={clsx(
				styles.root,
				{ [styles.firstOccurrence]: stack === -1 },
				className,
			)}>
			<BulletListContext.Provider
				value={stack + 1 >= bulletMap.length ? 0 : stack + 1}>
				{children}
			</BulletListContext.Provider>
		</Box>
	);
};
