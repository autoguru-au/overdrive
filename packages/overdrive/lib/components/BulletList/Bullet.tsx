import clsx from 'clsx';
import * as React from 'react';
import {
	FunctionComponent,
	isValidElement,
	ReactChild,
	useContext,
} from 'react';
import { useStyles } from 'react-treat';

import { Box } from '../Box';
import * as styleRefs from './Bullet.treat';
import { BulletList } from './BulletList';
import { BulletListContext, bulletMap, BulletType } from './context';

export interface Props {
	children: ReactChild;
	className?: string;
}

const getBulletCls = (styles: typeof styleRefs, type: BulletType): string => {
	switch (type) {
		case 'circle':
			return styles.root.circle;
		case 'square':
			return styles.root.square;
		case 'disc':
			return styles.root.disc;
		default:
			return '';
	}
};

export const Bullet: FunctionComponent<Props> = ({ children, className }) => {
	const styles = useStyles(styleRefs);

	return (
		<Box
			is="li"
			className={clsx(
				styles.root.default,
				getBulletCls(
					styles,
					bulletMap[useContext(BulletListContext)],
				) ?? '',
				{
					[styles.noDot]:
						isValidElement(children) &&
						children.type === BulletList,
				},
				className,
			)}>
			{children}
		</Box>
	);
};
