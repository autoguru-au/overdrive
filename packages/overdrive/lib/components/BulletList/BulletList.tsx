import clsx from 'clsx';
import * as React from 'react';
import { FunctionComponent, useContext } from 'react';
import { useStyles } from 'react-treat';

import { Box } from '../Box';
import * as styleRefs from './BulletList.treat';
import { BulletListContext, bulletMap } from './context';

export interface Props {
	className?: string;
}

export const BulletList: FunctionComponent<Props> = ({
	children,
	className,
}) => {
	const stack = useContext(BulletListContext);
	const styles = useStyles(styleRefs);

	return (
		<Box
			is="ul"
			className={clsx(className, styles.root, {
				[styles.firstOccurrence]: stack === -1,
			})}>
			<BulletListContext.Provider
				value={stack + 1 >= bulletMap.length ? 0 : stack + 1}>
				{children}
			</BulletListContext.Provider>
		</Box>
	);
};
