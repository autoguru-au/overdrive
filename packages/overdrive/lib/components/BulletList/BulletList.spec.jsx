import { render } from '@testing-library/react';
import * as React from 'react';

import { BulletList } from './BulletList';
import { BulletListContext } from './context';

describe('<BulletList />', () => {
	it('should not throw', () => {
		expect(() => render(<BulletList />)).not.toThrow();
	});

	it('should match the snapshot', () => {
		expect(render(<BulletList />).container.firstChild).toMatchSnapshot();
	});

	it.skip('should pass down className', () => {
		expect(
			render(<BulletList className="test-class" />).container.firstChild,
		).toHaveClass('test-class');
	});

	it.skip('should have the firstOccurrence class for the first and only first nested ul', () => {
		const { container } = render(
			<BulletList>
				<BulletList>
					<BulletList>test</BulletList>
				</BulletList>
			</BulletList>,
		);

		expect(container.querySelectorAll('.firstOccurrence').length).toBe(1);
	});

	it('should pass the correct nest value', () => {
		const validate = (expectation) => (
			<BulletListContext.Consumer>
				{(value) => {
					expect(value).toBe(expectation);
				}}
			</BulletListContext.Consumer>
		);

		render(
			<BulletList>
				{validate(0)}
				<BulletList>
					{validate(1)}
					<BulletList>{validate(2)}</BulletList>
				</BulletList>
			</BulletList>,
		);
	});
});
