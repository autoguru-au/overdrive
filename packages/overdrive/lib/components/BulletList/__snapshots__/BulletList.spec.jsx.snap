// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BulletList /> should match the snapshot 1`] = `
<ul
  class="reset_list__6swfds4 BulletList_root__1xhyplh0 BulletList_firstOccurrence__1xhyplh1"
/>
`;
